### ISSUE TITLE
Error on add to cart AJAX submission

### ISSUE THREAD

**Issue by sdussaut** [damcclean/simple-commerce#975]
### Description

Hi!

I am trying to add a product via AJAX as explained there (https://simple-commerce.duncanmcclean.com/ajax) and unfortunately I keep getting an error (419 status code, page expired, plain HTML response).

The form encryption is disabled (via `disable_form_parameter_validation` in the config).

### Steps to reproduce

1. Go to the product page
2. Do a simple POST request in JS after having prevented the submit on the form (I do it with a simple fetch() )
3. Payload is as follows:

```js
{
  "_token": "aG5JeI46QHsM2rjrqbkYpEeZZrqu9Zd9eqFmx9js",
  "_request": "Empty",
  "product": "9de4d1c5-c282-4199-9e1b-4bc82fa2a267",
  "quantity": "1"
}
```

I thought it could be related to the CSRF token being cached but with cleared caches and a fresh session the result is still the same.

Adding headers to the request yields mixed results:

- adding `X-CSRF-TOKEN` header gives a 500 error with `Exception: Unable to find Form Request`
- adding `'Content-Type': 'application/json'` header gives a 302 (?) followed by a 404 and a plain HTML response **but** adds the product to the cart
- adding `X-Requested-With` as per Statamic docs also leads to a 419 error but with a `CSRF token mismatch` message.

Any insight on what could be happening would be very helpful, thanks!

### Environment

Environment
Application Name: Test site
Laravel Version: 10.40.0
PHP Version: 8.3.1
Composer Version: 2.6.6
Environment: local
Debug Mode: ENABLED
URL: test-site.test
Maintenance Mode: OFF

Cache
Config: NOT CACHED
Events: NOT CACHED
Routes: NOT CACHED
Views: CACHED

Drivers
Broadcasting: log
Cache: statamic
Database: sqlite
Logs: stack / single
Mail: smtp
Queue: sync
Session: file

Simple Commerce
Currencies: EUR
Gateways: Dummy
Repository: Customer: DoubleThreeDigital\SimpleCommerce\Customers\EloquentCustomerRepository
Repository: Order: DoubleThreeDigital\SimpleCommerce\Orders\EloquentOrderRepository
Repository: Product: DoubleThreeDigital\SimpleCommerce\Products\EntryProductRepository
Shipping Methods: Free Shipping
Tax Engine: DoubleThreeDigital\SimpleCommerce\Tax\BasicTaxEngine

Statamic
Addons: 2
Antlers: runtime
Stache Watcher: Enabled
Static Caching: Disabled
Version: 4.44.0 PRO

Statamic Addons
doublethreedigital/runway: 5.6.1
doublethreedigital/simple-commerce: 5.9.1

**Comment by duncanmcclean** [comment]
Hey 👋

> adding X-CSRF-TOKEN header gives a 500 error with `Exception: Unable to find Form Request`

It looks like this fixes the 419 server error but you're now hitting a different error. Do you continue to see that error if you remove the `_request` parameter from the payload?

**Comment by sdussaut** [comment]
Yes, the same error occurs:

`Exception: Unable to find Form Request [http://test-site.test/products/simple-commerce-stickers] in file /test-site/vendor/doublethreedigital/simple-commerce/src/Http/Requests/AcceptsFormRequests.php on line 37`

The AJAX request is sent to the form action which is `http://test-site.test/!/simple-commerce/cart-items` but the form request here seems wrong. Shouldn't it be something like `/!/simple-commerce/cart-items` ?

**Comment by duncanmcclean** [comment]
I'm not quite sure why that error's happening or why it's using the previous URL as the form request. When I have some time, I'll try and replicate.

**Comment by duncanmcclean** [comment]
I've just had some time to try and re-create this and I've managed to get it working fine for me...

In my `simple-commerce.php` config file, I've got form parameter validation disabled:

```php
'disable_form_parameter_validation' => true,
```

Then, in my JavaScript, I've got `_request` set as `Empty`:

```js
axios.post(`/!/simple-commerce/cart-items`, {
    _token: document.querySelector('input[name="_token"]').value,
    _request: 'Empty',
    product: productId,
    quantity: 1,
})
```

Then, when I trigger the axios call, I get a successful response back:

![CleanShot 2024-01-20 at 16 59 32](https://github.com/duncanmcclean/simple-commerce/assets/19637309/d4c5eea7-ebcb-414d-acf4-d80d5f8e9c74)

I'm sure you've tried all of that though but might be worth giving it one more try. If you still can't get it working, maybe try and see if you can replicate the issue in a fresh Simple Commerce site.

**Comment by sdussaut** [comment]
Hello!
Thanks for the answer, I found what is happening.

I had already tried all that yes (even with a fresh repo) but with the same result in the end.
But what I had not tried (and your answer gave me a hint) was making a request **with axios** (I am using a regular `fetch`).

So I tried with it and it worked as expected. So I started to suspect axios was doing things differently behind the scenes. And… it did.

It is actually adding three headers to the request, two of them seem to be required for the request to be successful.

It adds this:

```
'Accept': 'application/json'
'Content-Type': 'application/json'
'X-XSRF-TOKEN': 'token value'
```

What was happening is that I was not specifying the `Accept` header and thus let the browser send its default value (which is `*/*`) and thus let the server *decide* what to send back as a response (this is the use case where the response is a 302 with HTML content).

Specifying both the `Accept` and `Content-Type` headers is enough for the calls to work and it makes sense :-)
The third header seems to be completely optional as the token is already sent in the request body.

Maybe adding a note about this in the docs (for people who do not use Axios) would be helpful?

Many thanks for your help, everything is fine now!
